### ISSUE TITLE
cryptsetup: RequiresMountFor dependency is ignored in initrd

### ISSUE THREAD

**Issue by nadiamoe** [systemd/systemd#13383]
**systemd version the issue has been seen with**
> 242.84

**Used distribution**
 > Arch Linux

**Expected behaviour you didn't see**
 cryptsetup-generator emits a `RequiresMountFor=` dependency for keyfiles. This (allegatedly) works on the real root, but not in initrd, at least when the device pointed by`RequiresMountFor=` also needs to be decrypted and mounted.

**Unexpected behaviour you saw**
 `systemd-cryptsetup@*` units do not find the keyfile and fall back to password input. I'd expect units with such dependency to wait until the required path is decrypted and mounted.

```
Aug 21 23:39:52 archlinux systemd[1]: Starting Cryptography Setup for smir2...
Aug 21 23:39:52 archlinux systemd-cryptsetup[308]: WARNING: Locking directory /run/cryptsetup is missing!
Aug 21 23:39:52 archlinux systemd-cryptsetup[308]: Set cipher aes, mode xts-plain64, key size 512 bits for device /dev/disk/by-partlabel/cryptosmir2.
Aug 21 23:39:52 archlinux systemd-cryptsetup[308]: Failed to open key file.
Aug 21 23:39:52 archlinux systemd-cryptsetup[308]: Failed to activate with key file '/mnt/keychain/example.key'. (Key file missing?)
Aug 21 23:40:14 archlinux systemd-cryptsetup[308]: Set cipher aes, mode xts-plain64, key size 512 bits for device /dev/disk/by-partlabel/cryptosmir2.
Aug 21 23:40:17 archlinux systemd[1]: Started Cryptography Setup for smir2.
```

Inspecting the initrd from the bultin emergency shell shows that the cryptsetup unit for `/dev/mapper/keychain`  is indeed successfully opened, but the opened device is not mounted. `/etc/fstab` is not present in the initrd, either.

**Steps to reproduce the problem**
Problem should be reproducible when using a crypttab like this:

```
keychain   PARTLABEL=encriptedkeychain # This must be decrypted with a password

mirror1    PARTLABEL=encryptedm1    /mnt/keychain/m1.key
mirror2    PARTLABEL=encryptedm2    /mnt/keychain/m2.key
# [...]
```

And of course, in fstab:
```
/dev/mapper/keychain  /mnt/keychain   ext4    defaults   0 0
```

**Notes**
I understand this is an unusual use case. If this is not supported on initrd, I'd happily understand. However it would be nice to have this limitation documented somewhere.

**Comment by poettering** [comment]
If you analyze the deps when this fails, what do you fine? i.e. if you look at "systemctl show" on the .mount and .device units, (in particular: Requires= and After= deps) do you see everything generated correctly?

It might be worth providing the generated unit files here from these crypttab/fstab entries. i.e. the stuff you find in /run/systemd/generator*/

**Comment by roobre** [comment]
Hello,

Thanks for the response. While I no longer have access to the same setup where I experienced this, I think I should be able to set up a VM reproducing the problem. It might take one or two days until I find the time tho. I apologize for this. I'll come back with the requested logs as soon as I can.

**Comment by roobre** [comment]
Hi again,

I got motivated and replicated the setup in a VM. I have targzd the contents of `/run/systemd/generator/`, you can find it [here](https://github.com/systemd/systemd/files/3828432/systemd-generator.tar.gz).

Since a few names have changed, here is the setup for the VM:

`/dev/vda1:` Boot, unencrypted
`/dev/vda2:` Keychain, encrypted ext4 with keyfiles for the other partitions
`/dev/vda3:` Root partition, encrypted ext4
`/dev/vda4:` Home partition, encrypted ext4

### `/etc/fstab`
```
# UUID=8f7b2fdd-2fcb-4c30-8377-8c77911a2705
/dev/mapper/root    	/         	ext4      	rw,relatime	0 1

# UUID=b4c7796f-b134-4866-b015-7b190be9c525
/dev/vda1           	/boot     	ext4      	rw,relatime	0 2

# UUID=83c65f0a-393b-4997-9643-315312dcbf77
/dev/mapper/keychain	/mnt/keychain	ext4      	rw,relatime	0 2

# UUID=cc392663-4986-48bb-a990-96a925a17bd5
/dev/mapper/home    	/home     	ext4      	rw,relatime	0 2
```

### `/etc/crypttab{,.initramfs}`
```
keychain	/dev/vda2	none
root		/dev/vda3	/mnt/keychain/master.key
home		/dev/vda4	/mnt/keychain/master.key
```

I expect systemd to ask me the password for `/dev/vda2` and then mount it and use the file stored there to unlock the other partitions. Instead, it asks the password for the three partitions. Then boot process finishes successfully.

I can also provide the `.qcow2` for this VM, if it helps, It's a minimal arch install with the setup described, weighting around 1.5GiB lzma-compressed.

Regards

**Comment by arvidjaar** [comment]
> cryptsetup-generator emits a `RequiresMountFor=` dependency for keyfiles ...  `/etc/fstab` is not present in the initrd, either.

`RequiresMountsFor` works only if corresponding mount units exist. In most common case these units are auto-generated from `/etc/fstab`, so if there is no `/etc/fstab` and units are not created by other means `RequiesMountsFor` is effectively no-op.

Note that modern systemd implicitly supports device specification in `/etc/crypttab` and generates corresponding mount unit (which seems to be missing from manual page except in example section). In your last example it would be
```
root /dev/vda3 master.key:/dev/vda2
root /dev/vda4 master.key:/dev/vda2
```

**Comment by roobre** [comment]
Hi again,

Thanks for your answer! This was indeed a step in the right direction. This mechanism sure lacks some documentation. I'll gladly make a PR improving the man page when the issue is solved.

Back to the issue, it seems that setting `root /dev/vda3 master.key:/dev/vda2` will not work, as systemd does not figure out that `/dev/vda2` is a LUKS that must be decrypted and fails saying "unknown filesystem crypto_LUKS". That's of course not a problem, because if I specify the path to the decrypted block dev it can figure it out nicely. So in the end it looks like this:

```
keychain	/dev/vda2	none
root		/dev/vda3	master.key:/dev/mapper/keychain
home		/dev/vda4	master.key:/dev/mapper/keychain
```

This seems to work fine, at first: I get asked for the password of the keychain (`vda2`), the filesystem gets mounted under `/run/systemd/cryptsetup/keydev-{root,home}` and then an attempt to open `vda3` and `vda4` with `master.key` is made. However, this attempt fails: the subsequent units (`systemd-cryptsetup@{root,home}`) exit with signal 9 due to... out of memory? [See full log here](https://0bin.net/paste/-R6JhPEpm+XTkAjf#RdBW9vxA3WesvOTpMPZvn4YP7iUiEiBUP9Ysomcduzf). The VM has 1GiB of RAM, which I understand should be way more than enough for this.

Honestly I'm at a loss here. I'm open to any suggestions about how to debug this further.

**Comment by poettering** [comment]
This is probably caused by the LUKS2 use of the Argon2 KDF which is memory hard and by default wants to use half your memory, hence if you have multiple devices to unlock at the same time, then this will trigger OOM.

They fixed this in cryptsetup 2.2 which we make use of in 408c81f62454684dfbff1c95ce3210d06f256e58

I guess that means all issues are dealt with? Closing hence.

**Comment by roobre** [comment]
Hi,

Thanks @poettering for the fix and reporting back!. I upgraded to systemd 244.1, and the decryption process now finishes without errors.

There is, however, a small catch: After switching to the real root, mounting the LUKS volume used to hold the keys (`/dev/mapper/keychain` in my example) as per its fstab entry fails, saying the device is already mounted. This is indeed true, as systemd mounts it under `/run/systemd/cryptsetup/keydev-$whatever`.
Perhaps systemd should unmount the volume after successful decryption? Or am I doing something wrong?

![DeepinScreenshot_select-area_20200121180801](https://user-images.githubusercontent.com/969721/72826687-95ca6480-3c79-11ea-828c-1859c030af94.png)

![DeepinScreenshot_select-area_20200121180842](https://user-images.githubusercontent.com/969721/72826685-9531ce00-3c79-11ea-9b76-8c69780e72ae.png)
